cpp/src/io/parquet/reader_impl_helpers.cpp

```diff
@@ -536,6 +554,304 @@ aggregate_reader_metadata::aggregate_reader_metadata(
+                      [&](auto const& idx) {
+                        return validate_schemas(arrow_schema_root.children[idx],
+                                                pq_schema_root.children_idx[idx]);
+                      })) {
+    CUDF_LOG_ERROR("Parquet reader encountered a mismatch between Parquet and arrow schema.",
+                   "arrow:schema not processed.");
+    return;
+  }
+
+  // All good, now co-walk schemas
+  std::for_each(thrust::make_counting_iterator(0),
+                thrust::make_counting_iterator(pq_schema_root.num_children),
+                [&](auto const& idx) {
+                  co_walk_schemas(arrow_schema_root.children[idx],
+                                  pq_schema_root.children_idx[idx]);
+                });
+}
+
+std::optional<std::string_view> aggregate_reader_metadata::decode_ipc_message(
+  std::string_view const serialized_message) const
+{
+  // Constants copied from arrow source and renamed to match the case
+  constexpr auto MESSAGE_DECODER_NEXT_REQUIRED_SIZE_INITIAL         = sizeof(int32_t);
+  constexpr auto MESSAGE_DECODER_NEXT_REQUIRED_SIZE_METADATA_LENGTH = sizeof(int32_t);
+  constexpr int32_t IPC_CONTINUATION_TOKEN                          = -1;
+
+  // message buffer
+  auto message_buf = serialized_message.data();
+  // current message (buffer) size
+  auto message_size = static_cast<int32_t>(serialized_message.size());
+
+  // Check for empty message
+  if (message_size == 0) {
+    CUDF_LOG_ERROR("Parquet reader encountered zero length arrow:schema.",
+                   "arrow:schema not processed.");
+    return std::nullopt;
+  }
+  // Check for improper message.
+  if (message_size - MESSAGE_DECODER_NEXT_REQUIRED_SIZE_INITIAL < 0) {
+    CUDF_LOG_ERROR("Parquet reader encountered unexpected arrow:schema message length.",
+                   "arrow:schema not processed.");
+  }
+
+  // Get the first 4 bytes (continuation) of the ipc message
+  int32_t continuation;
+  std::memcpy(&continuation, message_buf, MESSAGE_DECODER_NEXT_REQUIRED_SIZE_INITIAL);
+
+  // Check if the continuation matches the expected token
+  if (continuation != IPC_CONTINUATION_TOKEN) {
+    CUDF_LOG_ERROR("Parquet reader encountered unexpected IPC continuation token.",
+                   "arrow:schema not processed.");
+    return std::nullopt;
+  } else {
+    // Offset the message buf and reduce remaining size
+    message_buf += MESSAGE_DECODER_NEXT_REQUIRED_SIZE_INITIAL;
+    message_size -= MESSAGE_DECODER_NEXT_REQUIRED_SIZE_INITIAL;
+  }
+
+  // Check for improper message.
+  if (message_size - MESSAGE_DECODER_NEXT_REQUIRED_SIZE_METADATA_LENGTH < 0) {
+    CUDF_LOG_ERROR("Parquet reader encountered unexpected arrow:schema message length.",
+                   "arrow:schema not processed.");
+  }
+
+  // Get the next 4 bytes (metadata_len) of the ipc message
+  int32_t metadata_len;
+  std::memcpy(&metadata_len, message_buf, MESSAGE_DECODER_NEXT_REQUIRED_SIZE_METADATA_LENGTH);
```

Comment:
encapsulating this in a function that returns int32_t should reduce repetition here.
It could even update message_buf and message_size - if a following check fails, we don't really care if the pointer and size have been updated, we return empty object either way

Reply:
I was thinking of that at some point but the read may not always be `int32` if one of the constants change size in the future.

Reply:
I think for now it's safe to assume it will be `int32` so all done!